Tests/MongoSwiftTests/ConnectionStringTests.swift

```diff
@@ -441,4 +441,49 @@ final class ConnectionStringTests: MongoSwiftTestCase {
             options: MongoClientOptions(localThresholdMS: tooLarge)
         )).to(throwError(errorType: MongoError.InvalidArgumentError.self))
     }
+
+    func testCompressionOptions() throws {
+        // zlib level validation
+        expect(try Compressor.zlib(level: -2)).to(throwError(errorType: MongoError.InvalidArgumentError.self))
+        expect(try Compressor.zlib(level: 10)).to(throwError(errorType: MongoError.InvalidArgumentError.self))
+        expect(try Compressor.zlib(level: -1)).toNot(throwError())
+        expect(try Compressor.zlib(level: 9)).toNot(throwError())
+
+        // options are set correctly from options struct
+        var opts = MongoClientOptions()
+        opts.compressors = [.zlib]
+        var connStr = try ConnectionString("mongodb://localhost:27017", options: opts)
+        expect(connStr.compressors).to(equal(["zlib"]))
+
+        opts.compressors = [try .zlib(level: 6)]
+        connStr = try ConnectionString("mongodb://localhost:27017", options: opts)
+        expect(connStr.compressors).to(equal(["zlib"]))
+        expect(connStr.options?["zlibcompressionlevel"]?.int32Value).to(equal(6))
+
+        // options parsed correctly from string
+        connStr = try ConnectionString("mongodb://localhost:27017/?compressors=zlib&zlibcompressionlevel=6")
+        expect(connStr.compressors).to(equal(["zlib"]))
+        expect(connStr.options?["zlibcompressionlevel"]?.int32Value).to(equal(6))
+
+        // options struct overrides string
+        opts.compressors = []
+        connStr = try ConnectionString("mongodb://localhost:27017/?compressors=zlib", options: opts)
+        expect(connStr.compressors).to(beEmpty())
+
+        opts.compressors = [try .zlib(level: 6)]
+        connStr = try ConnectionString(
+            "mongodb://localhost:27017/?compressors=zlib&zlibcompressionlevel=4",
+            options: opts
+        )
+        expect(connStr.compressors).to(equal(["zlib"]))
+        expect(connStr.options?["zlibcompressionlevel"]?.int32Value).to(equal(6))
+
+        // duplicate compressors should error
+        opts.compressors = [.zlib, try .zlib(level: 1)]
+        expect(try ConnectionString("mongodb://localhost:27017", options: opts))
+            .to(throwError(errorType: MongoError.InvalidArgumentError.self))
+
+        // unfortunately, we can't error on an unsupported compressor provided via URI. libmongoc will generate a
```

Comment:
this is unfortunate. I suppose we could actually parse the URI ourselves and extract the `compressors=x` part if we wanted to in order to check this though. let me know what you think

Reply:
`mongoc_uri_get_compressors` returns just the valid ones then, right?

Reply:
correct

Reply:
Are there many other options like this where we can't verify? If not then I'm personally okay with just not verifying this one. If there are, then maybe some manual parsing of the connection string would be valuable.

Reply:
Let's defer to a later project where you develop a connection string parser in pure swift. I think the changes will be additive (including potentially throwing an error) right? So we can take our time to think it through more holistically there.

Reply:
I think under the philosophy that these should have been validated all along but aren't currently, you could consider the addition of such an error/validation to be a bug fix. 
but it is a bit breaking in the sense that if you were using an connection string containing an unsupported compressor before and relying on the driver's behavior to do that, it would no longer be the case.

I'm fine with considering the current inability to validate a bug and just opening a ticket to remember that we don't validate in this case right now. while completing the [validation ticket](https://jira.mongodb.org/browse/SWIFT-929) (last one left in the epic after this) it's possible I might encounter more cases that would require connection string parsing to properly validate, but I can't think of any off the top of my head that I've seen yet.

Reply:
FYI I opened [SWIFT-951](https://jira.mongodb.org/browse/SWIFT-951) to track that